File: app/src/main/java/com/devusercode/upchat/adapter/HomeAdapter.java
package com.devusercode.upchat.adapter;

import android.content.Intent;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.devusercode.upchat.ConversationActivity;
import com.devusercode.upchat.R;
import com.devusercode.upchat.models.User;
import com.devusercode.upchat.models.UserPair;
import com.devusercode.upchat.utils.ConversationUtil;
import com.google.android.material.card.MaterialCardView;

import java.util.List;

public class HomeAdapter extends RecyclerView.Adapter<HomeAdapter.HomeViewHolder> {
    private final String TAG = this.getClass().getSimpleName();
    private final List<UserPair> data;

    public HomeAdapter(List<UserPair> openConversations) {
        this.data = openConversations;
    }

    @NonNull
    @Override
    public HomeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.item_home_user, parent, false);
        return new HomeViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull HomeViewHolder holder, int position) {
        UserPair pair = data.get(position);

        holder.bind(pair);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    static class HomeViewHolder extends RecyclerView.ViewHolder {
        TextView username, last_message_text, last_message_time;
        MaterialCardView materialcardview1;
        ImageView profile_image;

        public HomeViewHolder(@NonNull View view) {
            super(view);

            materialcardview1 = view.findViewById(R.id.materialcardview1);
            profile_image = view.findViewById(R.id.profile_image);
            username = view.findViewById(R.id.username);
            last_message_text = view.findViewById(R.id.last_message_text);
            last_message_time = view.findViewById(R.id.last_message_time);
        }

        public void bind(UserPair pair) {
            User user = pair.getUser();
            String cid = pair.getConversationId();

            username.setText(user.getUsername());

            if (!user.getPhotoUrl().isEmpty()) {
                Glide.with(profile_image.getContext())
                        .load(Uri.parse(user.getPhotoUrl()))
                        .placeholder(R.drawable.ic_account_circle_black)
                        .circleCrop()
                        .into(profile_image);
            } else {
                profile_image.setImageResource(R.drawable.ic_account_circle_black);
            }

            ConversationUtil.getLastMessage(cid, last_msg -> {
                if (last_msg == null) {
                    last_message_text.setText("No messages available.");
                    last_message_time.setVisibility(View.GONE);

                } else if (last_msg.getSenderId().equals("system")) {
                    last_message_text.setVisibility(View.GONE);
                    last_message_time.setVisibility(View.GONE);

                } else {
                    last_message_text.setText(last_msg.getMessage());
                    last_message_time.setText(last_msg.getParsedTime());
                }
            });

            materialcardview1.setOnClickListener(view -> {
                Intent intent = new Intent(view.getContext(), ConversationActivity.class);
                intent.putExtra("uid", user.getUid());
                view.getContext().startActivity(intent);
            });
        }
    }
}

File: app/src/main/java/com/devusercode/upchat/adapter/HomeAdapter.kt
package com.devusercode.upchat.adapter

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.devusercode.upchat.ConversationActivity
import com.devusercode.upchat.R
import com.devusercode.upchat.models.UserPair
import com.devusercode.upchat.security.AES
import com.devusercode.upchat.utils.ConversationUtil
import com.google.android.material.card.MaterialCardView
import androidx.core.net.toUri

@RequiresApi(Build.VERSION_CODES.O)
class HomeAdapter(private var data: List<UserPair>) : RecyclerView.Adapter<HomeAdapter.HomeViewHolder>() {
    private val TAG = "HomeAdapter"

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HomeViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.item_home_user, parent, false)
        return HomeViewHolder(view)
    }

    override fun onBindViewHolder(holder: HomeViewHolder, position: Int) {
        val pair = data[position]
        holder.bind(pair)
    }

    override fun getItemCount(): Int {
        return data.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun update(newData: List<UserPair>) {
        data = newData
        notifyDataSetChanged()
    }

    class HomeViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var username: TextView = view.findViewById(R.id.username)
        private var lastMessageText: TextView = view.findViewById(R.id.last_message_text)
        private var lastMessageTime: TextView = view.findViewById(R.id.last_message_time)
        private var materialCardView: MaterialCardView = view.findViewById(R.id.materialcardview1)
        private var profileImage: ImageView = view.findViewById(R.id.profile_image)
        private var onlineStatus: ImageView = view.findViewById(R.id.online_status)

        @RequiresApi(Build.VERSION_CODES.O)
        fun bind(pair: UserPair) {
            val user = pair.user
            val cid = pair.conversationId

            username.text = user.username

            if (user.online.toBoolean()) {
                onlineStatus.setImageResource(R.drawable.ic_status_online)
            } else {
                onlineStatus.setImageResource(R.drawable.ic_status_offline)
            }

            if (user.photoUrl!!.isNotEmpty()) {
                Glide.with(profileImage.context)
                    .load(user.photoUrl?.toUri())
                    .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                    .placeholder(R.drawable.ic_account_circle_black)
                    .circleCrop()
                    .into(profileImage)
            } else {
                profileImage.setImageResource(R.drawable.ic_account_circle_black)
            }

            ConversationUtil.getLastMessage(cid) { lastMsg ->
                if (lastMsg == null) {
                    lastMessageText.visibility = View.GONE
                    lastMessageTime.visibility = View.GONE
                    return@getLastMessage
                }

                if (lastMsg.senderId.equals("system")) {
                    lastMessageText.visibility = View.GONE
                    lastMessageTime.visibility = View.GONE

                } else if (lastMsg.type == "image") {
                    lastMessageText.text = username.context.getString(R.string.home_last_message_photo)
                    lastMessageTime.text = lastMsg.parsedTime
                }
            }

            materialCardView.setOnClickListener { view: View ->
                val intent = Intent(view.context, ConversationActivity::class.java)
                intent.putExtra("uid", user.uid)
                view.context.startActivity(intent)
            }
        }
    }
}